vslm/global_object_manager.go

```diff
@@ -0,0 +1,659 @@
+	req := types.VslmRetrieveVStorageObject{
+		This: this.Reference(),
+		Id:   id,
+	}
+
+	res, err := methods.VslmRetrieveVStorageObject(ctx, this.c, &req)
+	if err != nil {
+		return nil, err
+	}
+
+	return &res.Returnval, nil
+}
+
+func (this *GlobalObjectManager) RetrieveState(ct context.Context, id vim.ID) (*vim.VStorageObjectStateInfo, error) {
+	req := types.VslmRetrieveVStorageObjectState{
+		This: this.Reference(),
+		Id:   id,
+	}
+
+	res, err := methods.VslmRetrieveVStorageObjectState(ct, this.c, &req)
+	if err != nil {
+		return nil, err
+	}
+
+	return &res.Returnval, nil
+}
+
+func (this *GlobalObjectManager) RetrieveAssociations(ct context.Context, ids []vim.ID) (
+	[]types.VslmVsoVStorageObjectAssociations, error) {
+	req := types.VslmRetrieveVStorageObjectAssociations{
+		This: this.Reference(),
+		Ids:  ids,
+	}
+
+	res, err := methods.VslmRetrieveVStorageObjectAssociations(ct, this.c, &req)
+	if err != nil {
+		return nil, err
+	}
+
+	return res.Returnval, nil
+}
+
+func (this *GlobalObjectManager) ListObjectsForSpec(ctx context.Context, query []types.VslmVsoVStorageObjectQuerySpec,
+	maxResult int32) (*types.VslmVsoVStorageObjectQueryResult, error) {
+	req := types.VslmListVStorageObjectForSpec{
+		This:      this.Reference(),
+		Query:     query,
+		MaxResult: maxResult,
+	}
+
+	res, err := methods.VslmListVStorageObjectForSpec(ctx, this.c, &req)
+	if err != nil {
+		return nil, err
+	}
+
+	return res.Returnval, nil
+}
+
+func (this *GlobalObjectManager) Clone(ctx context.Context, id vim.ID, spec vim.VslmCloneSpec) (*Task, error) {
+	req := types.VslmCloneVStorageObject_Task{
+		This: this.Reference(),
+		Id:   id,
+		Spec: spec,
+	}
+
+	res, err := methods.VslmCloneVStorageObject_Task(ctx, this.c, &req)
+	if err != nil {
+		return nil, err
+	}
+
+	return NewTask(this.c, res.Returnval), nil
+}
+
+func (this *GlobalObjectManager) Relocate(ct context.Context, id vim.ID, spec vim.VslmRelocateSpec) (*Task, error) {
+	req := types.VslmRelocateVStorageObject_Task{
+		This: this.Reference(),
+		Id:   id,
+		Spec: spec,
+	}
+
+	res, err := methods.VslmRelocateVStorageObject_Task(ct, this.c, &req)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return NewTask(this.c, res.Returnval), nil
+}
+
+func (this *GlobalObjectManager) SetControlFlags(ct context.Context, controlFlags []string) error {
+	req := types.VslmSetVStorageObjectControlFlags{
+		This:         this.Reference(),
+		ControlFlags: controlFlags,
+	}
+
+	_, err := methods.VslmSetVStorageObjectControlFlags(ct, this.c, &req)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (this *GlobalObjectManager) ClearControlFlags(ct context.Context) error {
+	req := types.VslmClearVStorageObjectControlFlags{
+		This: this.Reference(),
+	}
+
+	_, err := methods.VslmClearVStorageObjectControlFlags(ct, this.c, &req)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (this *GlobalObjectManager) AttachTag(ctx context.Context, id vim.ID, category string, tag string) error {
```

Comment:
`ObjectManager` uses `types.VslmTagEntry` for category+tag, since `ListAttachedTags` returns that type:

https://github.com/vmware/govmomi/blob/35d0b7d3d2f6ddb02af246d2e0285e5c23aa9850/vslm/object_manager.go#L360

`ObjectManager` also uses `id string` everywhere, rather than `vim.ID`.

I'd be open to changing `ObjectManager`'s method signatures if you prefer, but would like to keep them consistent.

Reply:
I'd prefer to use the ID object rather than a string.  For example, ListObjectsForSpec returns a VslmVsoVStorageObjectQuerySpec which has an array of IDs.  In order to retrieve an object, you'd have to convert the ID to a string and we'll then convert back to an ID.  Can we change those signatures in ObjectManager?  How much stuff will we break?

Reply:
I'd be ok with that, it'll break a bunch of the govc disk commands, but those are easy to fix.  It'll also break some of the K8s CSI driver for vSphere, but that won't be much trouble to fix either.  I don't know of anything else using the current `ObjectManager` APIs and they haven't been around too long.

Reply:
I'll update ObjectManager and our examples and make a separate PR